File: MATLAB/example_27/EnKF_storage.m
function S = ENKF_Storage(pars,plugin)
%<><><><><><><><><><><><><><><><><><><><><><><><><><><><><><><><><><><><><>
% Single layer interception model. After:
% Vrugt, J. A., S.C. Dekker, and W. Bouten, Identification of rainfall 
%   interception model parameters from measurements of throughfall and 
%   forest canopy storage, Water Resources Research, 39(9), 1251,
%   doi:10.1029/2003WR002013, 2003.          
% 
% SODA method. After:
% Vrugt, J.A., C.G.H. Diks, H.V. Gupta, W. Bouten, and J.M. Verstraten
%   (2005), Improved treatment of uncertainty in hydrologic modeling: 
%   Combining the strengths of global optimization and data assimilation, 
%   Water Resources Research, 41, W01017, doi:10.1029/2004WR003059
% © Jasper A. Vrugt, University of Amsterdam, 2003
%<><><><><><><><><><><><><><><><><><><><><><><><><><><><><><><><><><><><><>

persistent m N R C T P Ep alpha
if isempty(m)
    m = plugin.m; N = plugin.N; R = plugin.R; C = plugin.C;
    T = plugin.obsTime; P = plugin.Precip; Ep = plugin.Potevap;
    alpha = 0.05; % significance level (for EnKF prediction intervals)
end

% Data assimilation of the storage model using the Ensemble Kalman Filter.
% The parameters of this model are assumed known and their values are
% provided as input to this function, along with other variables that are
% stored in the structure "plugin". This variable allows user to port
% information from example_27 to this function.

% I recommend that you check the following lecture to understand in more 
% detail the concept, theory, and implementation of data assimilation: 
% https://www.youtube.com/watch?v=l5cb4ONIB-c&feature=youtu.be&disable_polymer=true
% including methods such as SODA. 
% This and other lectures can also be found on 
% my website: faculty.sites.uci.edu/jasper/teaching
%
% In short: SODA uses an inner layer of state estimation with the ENKF 
% embedded within an outer layer of MCMC simulation with DREAM. This 
% approach I originally developed to get a much better understanding of
% epistemic (model structural) errors - and to minimize the impact of input
% data and epistemic errors on the optimized parameter values. What is
% more, detailed analysis of the state updates can reveal new insights 
% into model strucxtural errors

% First unpack the parameter values of the Lorenz model
p = num2cell(pars); [a,b,c,d,AR_rho] = deal(p{:});

% Interception model equations in function InterceptionModelEquations
% This function is called by interception_Model --> using ode45 solver

%<><><><><><><><><><><><><><><><><><><><><><><><><><><><><><><><><><><><><>
%               Now execute ensemble Kalman Filter
%<><><><><><><><><><><><><><><><><><><><><><><><><><><><><><><><><><><><><>

% Set the first mean simulated value equal to observed value
S(1,1:m) = plugin.Y(1:m,1);
% Draw m x N matrix of measurement errors from N_m(0,R)
Meas_err = mvnrnd(zeros(1,m),R,N)';
% m x N matrix of initial states of ensemble members: min of zero
X_a = bsxfun(@plus,plugin.Y(1:m,1),Meas_err);
% m x N matrix of state forecasts of ensemble members
X_f = nan(m,N);
% Draw m x N matrix with model errors
Mod_err = mvnrnd(zeros(1,m),C,N)';

% Now run the model one-observation ahead
for t = 1 : plugin.NT - 1     % plugin.NT - 1 --> we start at t = 0
    % Generate state forecast for each ensemble member
    for ii = 1 : plugin.N
        % Run the model from one time step to next for the nth state vector
        [~,state_forecast] = ode45('InterceptionModelEquations',...
            [T(t) T(t+1)],X_a(1:m,ii),[],P(t,1),Ep(t,1),a,b,c,d);
        % Store the simulated state forecast at final time
        X_f(1:m,ii) = state_forecast(end,1:m);
        % Compute m x n matrix of model errors at time t using AR(1) scheme
        Mod_err(1:m,ii) = AR_rho * Mod_err(1:m,ii) + ...
            sqrt(1 - AR_rho^2) * mvnrnd(zeros(1,m),C)';
    end
    % m x N matrix of state forecasts
    X_f = max(X_f + Mod_err,0);
    % Compute m x 1 vector of mean state forecast and store in matrix E
    e = mean(X_f,2); S(t+1,1:m) = e';
    % Draw m x N matrix of measurement errors at time t
    Meas_err = mvnrnd(zeros(1,m),R,N)';
    % Generate m x N matrix of measured data replicates at time t + 1 
    D = bsxfun(@plus,plugin.Y(1:m,t+1),Meas_err);   
    %<><><><><><><><><><><><><><><><><><><><><><><><><><><><><><><><><><><>
    % Intermittent block (not needed for filter but for plotting)
    %<><><><><><><><><><><><><><><><><><><><><><><><><><><><><><><><><><><>
    
    % Now determine 95% ranges of forecasted state
    % T = sort(X_f,2); a1 = ceil(N*(alpha/2)); a2 = N - a1;
    % 2.5/97.5 percentile (column m+1 to 3*m of S)
    % S(t+1,m+1:2*m) = T(1:m,a1)'; S(t+1,2*m+1:3*m) = T(1:m,a2)';
    
    %<><><><><><><><><><><><><><><><><><><><><><><><><><><><><><><><><><><>
    % End intermittent block (not needed for filter but for plotting)
    %<><><><><><><><><><><><><><><><><><><><><><><><><><><><><><><><><><><>
    
    % Step 3: Subtract mean state from each vector (ensemble member) of Xf
    A = bsxfun(@minus,X_f,e);
    % Step 4: Compute m x m sample covariance matrix C of state forecasts
    C = 1/(N-1) * (A*A'); % which is similar to C = cov(A');
    % Step 5: Compute the m x m Kalman gain matrix 
    % K = C * inv(C + R); 
    K = C/(C+R);
    % Step 6: Compute m x N matrix of analysis states
    X_a = X_f + K * ( D - X_f );   
end

% Return a single vector of state variables (= storage) back to DREAM
S = S(:);

%<><><><><><><><><><><><><><><><><><><><><><><><><><><><><><><><><><><><><>
%                   End of ensemble Kalman Filter
%<><><><><><><><><><><><><><><><><><><><><><><><><><><><><><><><><><><><><>

end

File: Python/example_27/EnKF_storage.py
import numpy as np
from scipy.integrate import solve_ivp

############################
### Case study 27
############################
def EnKF_storage(par, plugin):
    # Vrugt, J. A., S.C. Dekker, and W. Bouten, Identification of rainfall 
    #   interception model parameters from measurements of throughfall and 
    #   forest canopy storage, Water Resources Research, 39(9), 1251,
    #   doi:10.1029/2003WR002013, 2003.   
    # Vrugt, J.A., C.G.H. Diks, H.V. Gupta, W. Bouten, and J.M. Verstraten
    #   (2005), Improved treatment of uncertainty in hydrologic modeling: 
    #   Combining the strengths of global optimization and data assimilation, 
    #   Water Resources Research, 41, W01017, doi:10.1029/2004WR003059
    if not hasattr(EnKF_storage, "initialized"):    # Store local variables in memory
        EnKF_storage.N = plugin['N']                # ensemble size EnKF
        EnKF_storage.m = plugin['m']                # number of measurements at given time
        EnKF_storage.R = plugin['R']                # m x m measurement error covariance matrix
        EnKF_storage.C = plugin['C']                # m x m model error covariance matrix at given time
        EnKF_storage.obsTime = plugin['obsTime']    # nx1 vector of observation times
        EnKF_storage.Precip = plugin['Precip']      # nx1 vector of measured precipitation rates: mm/d
        EnKF_storage.Potevap = plugin['Potevap']    # nx1 vector of measured potential evapo rates: mm/d
        def InterceptionModelEquations(t, S, P_int, E0_int, a, b, c, d):
            """
            Interception model equations.
            Args:
            - t: Current time
            - S: Current storage
            - P_int: Interpolated rainfall value at time t
            - E0_int: Interpolated potential evapotranspiration value at time t
            - a, b, c, d: Model parameters
            Returns:
            - dSdt: Rate of change of storage (mm/day)
            """
            I = a * P_int                       # Interception in mm/day
            if S > c:                           # Calculate drainage (only if storage is larger than storage capacity)
                D = b * (S - c)                 # in mm/day
            else:
                D = 0

            E = d * E0_int * S / c              # Calculate evaporation in mm/day
            dSdt = I - D - E                    # Now calculate the change in storage in mm/day

            return dSdt
   
        EnKF_storage.IMEquations = InterceptionModelEquations
        EnKF_storage.initialized = True         # Flag to indicate that initialization is comple

    a, b, c, d, AR_rho = par                    # Extract parameters from 'par'
    
    N = EnKF_storage.N
    m = EnKF_storage.m
    R = EnKF_storage.R
    C = EnKF_storage.C
    if np.isscalar(C):
        C = np.array([[C]])
    T = EnKF_storage.obsTime
    P = EnKF_storage.Precip
    Ep = EnKF_storage.Potevap

    S = np.full((plugin['NT'], m), np.nan)                              # Initialize matrix of storages
    S[0, :] = plugin['Y'][0]                                            # Set the first mean simulated value equal to observed value
    Meas_err = np.random.multivariate_normal(np.zeros(m), R, N).T       # Draw m x N matrix of measurement errors from N_m(0, R)
    X_a = np.add(plugin['Y'][0], Meas_err)                              # m x N matrix of initial states of ensemble members: min of zero
    X_f = np.nan * np.zeros((m, N))                                     # m x N matrix of state forecasts of ensemble members
    Mod_err = np.random.multivariate_normal(np.zeros(m), C, N).T        # Draw m x N matrix with model errors
    for t in range(plugin['NT'] - 1):                                   # Now run the model one-observation ahead
        for ii in range(N):                                             # Generate state forecast for each ensemble member
            # Solve the ODE using solve_ivp
            result = solve_ivp(EnKF_storage.IMEquations, [T[t], \
                T[t + 1]], X_a[0:m, ii], \
                args = (P[t],Ep[t],a,b,c,d), \
                method = 'RK45', t_eval = [T[t], T[t+1]])               # Run the model from one time step to next for the nth state vector
            state_forecast = result.y[:, -1]                            # Get the state forecast at the final time
            X_f[0:m, ii] = state_forecast                               # Store the simulated state forecast at final time
            # X_f[:m, ii] = state_forecast[-1, :m]
            Mod_err[:m, ii] = AR_rho * Mod_err[:m, ii] + \
                np.sqrt(1 - AR_rho**2) * np.random.multivariate_normal(np.zeros(m), C)   # Compute m x N matrix of model errors at time t using AR(1) scheme

        X_f += Mod_err                                                  # m x N matrix of state forecasts
        s = np.mean(X_f, axis = 1)                                      # Compute m x 1 vector of mean state forecast
        S[t + 1, :] = s                                                 # Store mean state forecast
        Meas_err = np.random.multivariate_normal(np.zeros(m), R, N).T   # Draw m x N matrix of measurement errors at time t
        D = np.add(plugin['Y'][t+1], Meas_err)                          # Generate m x N matrix of measured data replicates at time t + 1
        A = X_f - s[:, np.newaxis]                                      # Step 3: Subtract mean state from each vector (ensemble member) of Xf
        C = 1 / (N - 1) * np.dot(A, A.T)                                # Step 4: Compute m x m sample covariance matrix C of state forecasts
        K = np.linalg.solve(C + R, C)                                   # Step 5: Compute the m x m Kalman gain matrix
        X_a = X_f + np.dot(K, (D - X_f))                                # Step 6: Compute m x N matrix of analysis states
    
    s = S.flatten()                                                     # Return mean state forecast at each time as a single vector

    return s
